src/components/NanoContract/SelectAddressModal.component.js

```diff
@@ -0,0 +1,204 @@
+/**
+ * Copyright (c) Hathor Labs and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import React, { useEffect, useState } from 'react';
+import {
+  StyleSheet,
+  View,
+  FlatList,
+  TouchableHighlight,
+  Text,
+} from 'react-native';
+import { t } from 'ttag';
+import { useSelector } from 'react-redux';
+import { COLORS } from '../../styles/themes';
+import { ModalBase } from '../ModalBase';
+import { TextValue } from '../TextValue';
+import { TextLabel } from '../TextLabel';
+import { EditAddressModal } from './EditAddressModal.component';
+import { consumeAsyncIterator } from '../../utils';
+
+/**
+ * Use this modal to select an address from the wallet.
+ *
+ * @param {Object} props
+ * @param {string} props.address It refers to the address selected
+ * @param {boolean} props.show It determines if modal must show or hide
+ * @param {(address:string) => {}} props.onSelectAddress
+ * Callback function called when an address is selected
+ * @param {() => {}} props.onDismiss
+ * Callback function called to dismiss the modal
+ *
+ * @example
+ * <SelectAddressModal
+ *   address={selectedAddress}
+ *   show={showSelectAddressModal}
+ *   onDismiss={toggleSelectAddressModal}
+ *   onSelectAddress={handleSelectAddress}
+ * />
+ */
+export const SelectAddressModal = ({ address, show, onSelectAddress, onDismiss }) => {
+  const wallet = useSelector((state) => state.wallet);
+  const [addresses, setAddresses] = useState([]);
+  const [selectedItem, setSelectedItem] = useState({ address });
+  const [showEditAddressModal, setShowEditAddressModal] = useState(false);
+
+  const toggleEditAddressModal = () => {
+    setShowEditAddressModal(!showEditAddressModal);
+  };
+
+  const onDismissEditAddressModal = () => {
+    setSelectedItem({ address });
+    toggleEditAddressModal();
+  };
+
+  // This method is only called by AddressItem if the selected address
+  // is different from the current one.
+  const onSelectItem = (item) => {
+    setSelectedItem(item);
+    toggleEditAddressModal();
+  };
+
+  const hookAddressChange = (selectedAddress) => {
+    toggleEditAddressModal();
+    onSelectAddress(selectedAddress);
+  };
+
+  useEffect(() => {
+    const fetchData = async () => {
+      const iterator = await wallet.getAllAddresses();
+      const allAddresses = await consumeAsyncIterator(iterator);
+      setAddresses(allAddresses);
+    };
+    fetchData();
```

Comment:
Hummm... if we use a saga effect for this we would have to hold the address list in memory, but this is an ephemeral component and not all users will use it. My point is, it is not a core feature, it is more of a convenience. I like to think the saga effects as part of the wallet's feature cycles.

I can handle the errors using an utility `getAllAddresses` method that wraps the `wallet.getAllAddresses`.

Reply:
I've implemented the utility method and now the fetch is handling any possible error. The modal also have feedback content for when it is loading, in case the fetch request takes a while.

Reply:
>  My point is, it is not a core feature, it is more of a convenience. I like to think the saga effects as part of the wallet's feature cycles.

We added redux-saga to the project to deal with all side effects, the idea is to keep component as pure as possible and just reflect the app state stored in redux

So I disagree with handling the error inside the component

Reply:
I see now. Ok! I'll move it to saga.

Reply:
Refactored. See at 6432c23e75bdaa325d34466c3060540e344345a5